### ISSUE TITLE
[Bug] 

### ISSUE THREAD

**Issue by ice-tong** [pytorch/pytorch#115610]
### 🐛 Describe the bug

There is a big difference between the output of Conv float precision and double precision in torch1.12+cu113.

The code sample for reproduce the issue:

```python
# pylint: disable=missing-module-docstring, not-callable, invalid-name, line-too-long, no-member
import torch
import torch.nn.functional as F

torch.backends.cudnn.benchmark = False
torch.backends.cudnn.deterministic = True


def run_conv_forward_with_given_dtype(dtype) -> None:
    """Run the conv forward with given dtype.

    The conv inputs are from a real model.
    """
    input_tensor = torch.load("Conv_4_input.pt").to(dtype)
    weight_tensor = torch.load("Conv_4_weight.pt").to(dtype)
    bias_tensor = torch.load("Conv_4_bias.pt").to(dtype)

    out = F.conv2d(
        input_tensor,
        weight_tensor,
        bias_tensor,
        stride=[1, 1],
        padding=[0, 0],
        dilation=[1, 1],
        groups=1,
    )
    return out


print(f"The PyTorch version: {torch.__version__}")
print(f"The CUDNN version: {torch.backends.cudnn.version()}")

out_float = run_conv_forward_with_given_dtype(torch.float32)
out_double = run_conv_forward_with_given_dtype(torch.float64)

absolute_error = (out_float - out_double).abs()
relative_error = (out_float - out_double).abs() / out_double
print(f"The absolute error between float32 and float64: {absolute_error.max().item()}")
print(f"The relative error between float32 and float64: {relative_error.max().item()}")

index = torch.argmax(absolute_error)
print(f"The `{index}` element of the float32 output: {out_float.flatten()[index].item()}")
print(f"The `{index}` element of the float64 output: {out_double.flatten()[index].item()}")
```

Got output:
```
The PyTorch version: 1.12.1+cu113
The CUDNN version: 8302
The absolute error between float32 and float64: 0.002784239179315584
The relative error between float32 and float64: 94.2813634888402
The `1156171` element of the float32 output: -4.7595906257629395
The `1156171` element of the float64 output: -4.762374864942255
```

**All the code and related file can be found at: [bug_repro.tar.gz](https://github.com/pytorch/pytorch/files/13643091/bug_repro.tar.gz)**

### Versions

N/A

**Comment by ice-tong** [comment]
Hi @malfet , thanks for your quick reply!

Here are the outputs of `python3 -mtorch.utils.collect_env`:

```
❯ python3 -mtorch.utils.collect_env
Collecting environment information...
PyTorch version: 1.12.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-121-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090
GPU 4: NVIDIA GeForce RTX 3090
GPU 5: NVIDIA GeForce RTX 3090
GPU 6: NVIDIA GeForce RTX 3090
GPU 7: NVIDIA GeForce RTX 3090

Nvidia driver version: 525.147.05
cuDNN version: Probably one of the following:
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.2.1
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] torch==1.12.1+cu113
[pip3] torchaudio==0.12.1+cu113
[pip3] torchvision==0.13.1+cu113
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] torch                     1.12.1+cu113             pypi_0    pypi
[conda] torchaudio                0.12.1+cu113             pypi_0    pypi
[conda] torchvision               0.13.1+cu113             pypi_0    pypi
```

**Comment by ice-tong** [comment]
> Also, is there a reason you can't try it with latest PyTorch + CUDA-11.8? (I vaguely remember numerical issues with CUDNN-8.3)

We recently upgraded our code repository from torch 1.8 to 1.12, and discovered that this issue causes serious model accuracy problems. 

We wanted to use a slightly more stable version, so we didn't use the latest PyTorch + CUDA version. ^_^

**Comment by eqy** [comment]
I'm a bit confused as to how you were able to use a CUDA 10.2 build of PyTorch with RTX 3090, which is compute capability 8.6 when the max compute capability supported by CUDA 10.2 is 7.5.

Note that the 3090 supports TF32 in the hardware and it would be enabled by default. Unless you can show that "allclose" tests fail with this shape and expected tolerances (TF32 has the same precision as fp16), this would be expected behavior. You can set `torch.backends.cudnn.allow_tf32 = False` to disable TF32 to increase precision at the cost of performance.
